### ISSUE TITLE
Support do-notation patterns in custom hint rules (e.g. `_ <- f x` → `f_ x`)

### ISSUE THREAD

**Issue by NorfairKing** [ndmitchell/hlint#1695]
## Problem

I'd like to write a custom hlint rule that suggests replacing `_ <- DB.insert x` with `DB.insert_ x` in do-notation blocks. This is a common pattern with Persistent's `insert`/`insert_` and similar paired functions.

Currently, custom hint rules only support expression-level patterns using operators like `>>=` and `=<<`:

```yaml
- warn: {lhs: "concat (map f x)", rhs: "concatMap f x"}
```

There's no way to match on do-notation statements. A rule like this doesn't work:

```yaml
- warn: {lhs: "_ <- DB.insert x", rhs: "DB.insert_ x"}
```

## Desired behavior

Support matching do-notation statement patterns in `lhs`/`rhs`, at minimum the `_ <- f x` pattern. For example:

```yaml
- warn: {lhs: "_ <- DB.insert x", rhs: "DB.insert_ x", name: "Use DB.insert_"}
```

This would flag code like:

```haskell
do
  _ <- DB.insert record
  pure ()
```

And suggest:

```haskell
do
  DB.insert_ record
  pure ()
```

## Motivation

Many libraries provide paired functions where one returns a value and the other discards it (e.g. Persistent's `insert`/`insert_`, `insertMany`/`insertMany_`, `delete`/`delete_`). Binding the result to `_` is a code smell that's easy to catch mechanically but tedious to enforce via code review. HLint's existing `void` rule (`_ <- x` → `void x`) is similar in spirit but doesn't help when there's a dedicated discarding variant.
